### PR TITLE
Freshli.io lead gen MVP

### PR DESCRIPTION
**Pull request by mscottford**
Adds super simple start at a website for Freshli.io. On the "homepage" users can provide a URL to be scanned. Calculation work is handled by a background job worker, and results are stored in the database used by Freshli.io.

There is still a lot more to do, but this is a pretty good point to merge into main at this point.

### CODE REVIEW

File: .github/workflows/dotnet-core.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
     - name: Setup .NET Core
       uses: actions/setup-dotnet@v1
       with:
-        dotnet-version: 3.1.300
+        dotnet-version: 3.1.301
```
Comment (nickiemc) [inline .github/workflows/dotnet-core.yml:19]: A very meta update 💯

File: Dockerfile.web [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+RUN apt-get update && apt-get install -y unzip
+
+COPY docker-entrypoint-web.sh .
+
+RUN echo 'export PATH="$PATH:/root/.dotnet/tools"' > /root/.bashrc && \
+    . /root/.bashrc && \
+    dotnet tool install --global dotnet-ef
+
+# copy csproj and restore as distinct layers
+# COPY *.sln .
```
Comment (nickiemc) [inline Dockerfile.web:15]: Delete commented line.

File: Dockerfile.worker [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# https://hub.docker.com/_/microsoft-dotnet-core
+FROM mcr.microsoft.com/dotnet/core/sdk:3.1-bionic AS build
+RUN dotnet --version
+WORKDIR /source
+
+RUN apt-get update && apt-get install -y unzip
+# RUN ln -s /usr/lib/x86_64-linux-gnu/libssl.so.1.1 /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0
+# RUN ln -s /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
```
Comment (nickiemc) [inline Dockerfile.worker:8]: Can these two lines be deleted? Are they needed for reference for some reason?

File: Freshli.Web.Worker/appsettings.Development.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft": "Warning",
+      "Microsoft.Hosting.Lifetime": "Information"
+    }
```
Comment (nickiemc) [inline Freshli.Web.Worker/appsettings.Development.json:7]: Should `Hangfire` be included here as well?

File: Freshli.Web/Controllers/HomeController.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Freshli.Web.Data;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using Freshli.Web.Models;
+using Hangfire;
```
Comment (nickiemc) [inline Freshli.Web/Controllers/HomeController.cs:10]: Delete unused `using` statements.

File: Freshli.Web/Models/ErrorViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+using System;
+
+namespace Freshli.Web.Models {
+  public class ErrorViewModel {
+    public string RequestId { get; set; }
+
+    public bool ShowRequestId => !string.IsNullOrEmpty(RequestId);
```
Comment (nickiemc) [inline Freshli.Web/Models/ErrorViewModel.cs:7]: Does `RequestIdIsSet` make more sense?

File: Freshli.Web/Program.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
```
Comment (nickiemc) [inline Freshli.Web/Program.cs:8]: Delete unused `using` statements.

File: Freshli.Web/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            pattern: "{controller=Home}/{action=Index}/{id?}"
+          );
+        }
+      );
+    }
+  }
+
+  public class HangfireAuthorizationFilter : IDashboardAuthorizationFilter {
+    public bool Authorize([NotNull] DashboardContext context) {
+      // TODO: !THIS _CANNOT_ GO TO PRODUCTION LIKE THIS!
```
Comment (nickiemc) [inline Freshli.Web/Startup.cs:91]: Can you create an issue to address this to make sure it doesn't in fact make it into production?

File: Freshli.Web/Views/Home/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    <h1 class="display-4">Welcome</h1>
+    <p>Learn about <a href="https://docs.microsoft.com/aspnet/core">building Web apps with ASP.NET Core</a>.</p>
+</div>
+
+<div>
+    <form asp-route="CreateAnalysisRequest" method="post">
+        <div>
+            <label asp-for="Url">Public Git URL:</label>
+            <input asp-for="Url" type="url"/>
+        </div>
```
Comment (nickiemc) [inline Freshli.Web/Views/Home/Index.cshtml:16]: This is nitpicking, but, unless the other two fields are optional, I'd move this after them.

File: Freshli.Web/Views/Home/Privacy.cshtml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿@{
+    ViewData["Title"] = "Privacy Policy";
+}
+<h1>@ViewData["Title"]</h1>
+
+<p>Use this page to detail your site's privacy policy.</p>
```
Comment (nickiemc) [inline Freshli.Web/Views/Home/Privacy.cshtml:6]: Can you create an issue to update this page?

File: Freshli.Web/Views/Shared/Error.cshtml [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+<h3>Development Mode</h3>
+<p>
+    Swapping to <strong>Development</strong> environment will display more detailed information about the error that occurred.
+</p>
+<p>
+    <strong>The Development environment shouldn't be enabled for deployed applications.</strong>
+    It can result in displaying sensitive information from exceptions to end users.
+    For local debugging, enable the <strong>Development</strong> environment by setting the <strong>ASPNETCORE_ENVIRONMENT</strong> environment variable to <strong>Development</strong>
+    and restarting the app.
+</p>
```
Comment (nickiemc) [inline Freshli.Web/Views/Shared/Error.cshtml:25]: Do we want this to display in production? Do we need to add a flag so this isn't displayed in production?

File: Freshli.Web/Views/Shared/_Layout.cshtml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    </header>
+    <div class="container">
+        <main role="main" class="pb-3">
+            @RenderBody()
+        </main>
+    </div>
+
+    <footer class="border-top footer text-muted">
+        <div class="container">
+            &copy; 2020 - Freshli.Web - <a asp-area="" asp-controller="Home" asp-action="Privacy">Privacy</a>
```
Comment (nickiemc) [inline Freshli.Web/Views/Shared/_Layout.cshtml:40]: Update year to be dynamic or create an issue to do so.

File: Freshli.Web/appsettings.Development.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft": "Warning",
+      "Microsoft.Hosting.Lifetime": "Information"
+    }
```
Comment (nickiemc) [inline Freshli.Web/appsettings.Development.json:7]: Should `Hangfire` be included as well?

File: Freshli.Web/wwwroot/js/site.js [hunk 14]
```diff
@@ -0,0 +1,4 @@
+﻿// Please see documentation at https://docs.microsoft.com/aspnet/core/client-side/bundling-and-minification
+// for details on configuring this project to bundle and minify static web assets.
+
+// Write your JavaScript code.
```
Comment (nickiemc) [inline Freshli.Web/wwwroot/js/site.js:4]: Is this file required? Can it be deleted? If not can the comments be updated to include verbiage about why it's included?

Comment (mscottford) [reply Freshli.Web/wwwroot/js/site.js:4]: This was generated by the `dotnet new mvc` command which was used to initially create the `Freshli.Web` project.

File: .github/workflows/dotnet-core.yml (hunk 1 above)
Comment (mscottford) [reply .github/workflows/dotnet-core.yml:19]: 😄

File: Dockerfile.web (hunk 2 above)
Comment (mscottford) [reply Dockerfile.web]: Removed in f112bac.

File: Dockerfile.worker (hunk 3 above)
Comment (mscottford) [reply Dockerfile.worker]: Removed in f112bac.

File: Freshli.Web.Worker/appsettings.Development.json (hunk 4 above)
Comment (mscottford) [reply Freshli.Web.Worker/appsettings.Development.json:8]: I'm not sure. I added it just in case in e530c5d.

File: Freshli.Web/Controllers/HomeController.cs (hunk 5 above)
Comment (mscottford) [reply Freshli.Web/Controllers/HomeController.cs]: Fixed in cf6f5a4.

File: Freshli.Web/Models/ErrorViewModel.cs (hunk 6 above)
Comment (mscottford) [reply Freshli.Web/Models/ErrorViewModel.cs:5]: This was autogenerated by the `dotnet new mvc` command. I'm not a fan of naming boolean fields like that. Sounds like a method name. I'll fix this.

File: Freshli.Web/Program.cs (hunk 7 above)
Comment (mscottford) [reply Freshli.Web/Program.cs]: Fixed in cf6f5a4.

File: Freshli.Web/appsettings.Development.json (hunk 13 above)
Comment (mscottford) [reply Freshli.Web/appsettings.Development.json:8]: Fixed in e530c5d.

File: Freshli.Web/wwwroot/js/site.js (hunk 14 above)
Comment (mscottford) [reply Freshli.Web/wwwroot/js/site.js:4]: I'd like for us to switch to using npm or yarn for managing dependencies and using something from the node ecosystem to bundle our code for delivery. That would replace all of the JS that was autogenerated. I'll create an issue to take care of that.

Comment (mscottford) [reply Freshli.Web/wwwroot/js/site.js:4]: Created issue #55.

File: Freshli.Web/Views/Shared/_Layout.cshtml (hunk 12 above)
Comment (mscottford) [reply Freshli.Web/Views/Shared/_Layout.cshtml:40]: Created issue #56.

File: Freshli.Web/Views/Home/Privacy.cshtml (hunk 10 above)
Comment (mscottford) [reply Freshli.Web/Views/Home/Privacy.cshtml:6]: Created issue #57.

File: Freshli.Web/Startup.cs (hunk 8 above)
Comment (mscottford) [reply Freshli.Web/Startup.cs:91]: Created issue #58.

File: Freshli.Web/Views/Shared/Error.cshtml (hunk 11 above)
Comment (mscottford) [reply Freshli.Web/Views/Shared/Error.cshtml:25]: Created issue #59.

File: Freshli.Web/Models/ErrorViewModel.cs (hunk 6 above)
Comment (mscottford) [reply Freshli.Web/Models/ErrorViewModel.cs:5]: Created issue #59.

File: Freshli.Web/Views/Home/Index.cshtml (hunk 9 above)
Comment (mscottford) [reply Freshli.Web/Views/Home/Index.cshtml:16]: Created issue #60 to track getting this cleaned up and made more clear.

**Comment by mscottford** [comment]
@nickiemc I think that I have addressed all of your comments. Please take another peek and let me know.

**Comment by nickiemc** [comment]
I made some minor tweaks to the README. Everything else looks good!
